Add UDPClient and add Timer and Pipeline to StatsdClient

panpan-wu opened the pull request:
1. add UDPClient which reconnects automatically when error occurs
2. add Timer and Pipeline to StatsdClient

panpan-wu committed:
1. add UDPClient which reconnects automatically when error occurs
2. add Timer and Pipeline to StatsdClient

zeaphoo commented on basepy/asynclib/datagram.py:
```diff
@@ -187,6 +188,46 @@ def resume_writing(self):
         super().resume_writing()
 
 
+class UDPClient:
```
DatagramAutoClient or DatagramConnection ?

zeaphoo commented on basepy/asynclib/datagram.py:
```diff
@@ -187,6 +188,46 @@ def resume_writing(self):
+            try:
+                await self._datagram_client.send(data)
+            except Exception:
+                await self._init()
```
Normal should close the connection when exception on send, re-init the connection when next send called.

panpan-wu committed:
1. rename UDPClient to DatagramAutoClient
2. DatagramAutoClient: close connection when error occurs and reinit connection when next send called

zeaphoo commented on basepy/asynclib/datagram.py:
```diff
@@ -199,34 +199,44 @@ def __init__(self, host, port):
         self._last_connect_time = None
 
     async def init(self):
```
seems not useful?

panpan-wu replied:
init 没有忽略异常，初衷是如果 host 有问题，让外界初始化时能发现，比如 ConnectionRefusedError。而 send 中的初始化是忽略了所有异常，调用 send 不管有没有错误都不会抛出。

zeaphoo replied:
有可能会忘记调用这个函数，如果希望做到这个设计，也可以_init(_get_client)里实现。

zeaphoo commented on basepy/asynclib/datagram.py:
```diff
@@ -199,34 +199,44 @@ def __init__(self, host, port):
+        self._last_connect_time = int(time.time())
+        self._datagram_client = await connect(self._addr)
+
+    async def _init(self):
```
```suggestion
    async def _get_client(self):
        if self._datagram_client:
            return self._datagram_client
        not_init = self._last_connect_time is None
        waited_longer = int(time.time()) - self._last_connect_time) >= self.RECONNECT_INTERVAL
        if (not_init or waited_longer):
            try:
                await self._connect()
                return self._datagram_client
            except Exception:
                pass
        return None
```

zeaphoo commented on basepy/asynclib/datagram.py:
```diff
@@ -199,34 +199,44 @@ def __init__(self, host, port):
 
     write = send
 
+    async def _connect(self):
```
```suggestion
    async def _connect(self):
        self._datagram_client = await connect(self._addr)
        self._last_connect_time = int(time.time())
```

panpan-wu replied:
如果先设置 datagram_client 再设置 last_connect_time，last_connect_time 可能会一直为 None（在 connect 失败时），这样 RECONNECT_INTERVAL 就会失效。相关代码：

not_init = self._last_connect_time is None
waited_longer = int(time.time()) - self._last_connect_time) >= self.RECONNECT_INTERVAL
if (not_init or waited_longer):
    try:
        await self._connect()
        return self._datagram_client
    except Exception:
        pass

zeaphoo replied:
这个是我理解错了，_last_connect_time 表示的是去连接的时间，不是连接成功的时间。

zeaphoo commented on basepy/asynclib/datagram.py:
```diff
@@ -199,34 +199,44 @@ def __init__(self, host, port):
-            except Exception:
-                pass
-
+        await self._init()
```
```suggestion
        client = await self._get_client()
        if client is not None:
            try:
                await client.send(data)
            except:
                self.close()
```

panpan-wu committed:
DatagramAutoClient:
    1. change reconnect_interval from s to ms
    2. record send and connect exception
